docs/src/keymap-upgrade/textedit.ts

```diff
@@ -0,0 +1,182 @@
+        startIndex = node.startIndex;
+        endIndex = node.endIndex;
+      } else {
+        throw new TypeError();
+      }
+    } else if (typeof endIndex !== "number" || typeof newText !== "string") {
+      throw new TypeError();
+    }
+
+    this.startIndex = startIndex;
+    this.endIndex = endIndex;
+    this.newText = newText;
+  }
+}
+
+export type MatchFunc = (node: SyntaxNode, text: string) => boolean;
+export type ReplaceFunc = (
+  node: SyntaxNode,
+  replacement: string | null
+) => TextEdit[];
+
+/**
+ * Gets a list of text edits to apply based on a node and a map of text
+ * replacements.
+ *
+ * If replaceHandler is given, it will be called if the node matches a
+ * deprecated value and it should return the text edits to apply.
+ * Otherwise, the full node is replaced by the new text.
+ *
+ * If isMatch is given, it will be called to check if a node matches a
+ * deprecated value. Otherwise, the node's text is matched against the
+ * deprecated text.
+ *
+ * @param {SyntaxNode} node
+ * @param {Record<string, string | null>} replacementMap
+ * @param {ReplaceFunc} [replaceHandler]
+ * @param {MatchFunc} [isMatch]
+ */
+export function getUpgradeEdits(
+  node: SyntaxNode,
+  replacementMap: Record<string, string | null>,
+  replaceHandler?: ReplaceFunc,
+  isMatch?: MatchFunc
+) {
+  const defaultReplace: ReplaceFunc = (node, replacement) => [
+    new TextEdit(node, replacement ?? ""),
+  ];
+  const defaultMatch: MatchFunc = (node, text) => node.text === text;
+
+  replaceHandler = replaceHandler ?? defaultReplace;
+  isMatch = isMatch ?? defaultMatch;
+
+  for (const [deprecated, replacement] of Object.entries(replacementMap)) {
+    if (isMatch(node, deprecated)) {
+      return replaceHandler(node, replacement);
+    }
+  }
+  return [];
+}
+
+/**
+ * Sorts a list of text edits in ascending order by position.
+ */
+function sortEdits(edits: TextEdit[]) {
+  return edits.sort((a, b) => a.startIndex - b.startIndex);
+}
+
+export type Range = [number, number];
+
+export interface EditResult {
+  text: string;
+  changedRanges: Range[];
+}
+
+interface TextChunk {
+  text: string;
+  changed?: boolean;
+}
+
+/**
+ * Returns a string with text replacements applied.
+ */
+export function applyEdits(text: string, edits: TextEdit[]) {
+  // If we are removing text and it's the only thing on a line, remove the whole line.
+  edits = edits.map((e) => (e.newText ? e : expandEditToLine(text, e)));
+  edits = sortEdits(edits);
+
+  const chunks: TextChunk[] = [];
+  let currentIndex = 0;
+
+  for (let edit of edits) {
+    if (edit.startIndex < currentIndex) {
+      console.warn("discarding overlapping edit", edit);
+      continue;
+    }
+
+    chunks.push({ text: text.substring(currentIndex, edit.startIndex) });
+    chunks.push({ text: edit.newText, changed: true });
+    currentIndex = edit.endIndex;
+  }
+
+  chunks.push({ text: text.substring(currentIndex) });
+
+  return chunks.reduce<EditResult>(
```

Comment:
Perhaps a comment explaining this reduce would be helpful.

Reply:
Done. I also pulled some more of this code out into a separate function to simplify it.